### ISSUE TITLE
CustomResources allow GC finalizers

### ISSUE THREAD

**Issue by deads2k** [kubernetes/kubernetes#47431]
CustomResources enable GC when it is enabled overall, but the GC controller doesn't locate CRs, so the finalizer is never cleared, so the CR is never deleted.

@kubernetes/sig-api-machinery-misc @enisoc @ironcladlou

**Comment by enisoc** [comment]
I changed "fixes" to "xref" on the PR because I'd like to keep this issue open until we have a regression test. I can work on adding one if you won't get to it.

**Comment by enisoc** [comment]
Do you have repro steps for this?

**Comment by deads2k** [comment]
> Do you have repro steps for this?

```
[deads@deads-dev-01 kubernetes]$ kubectl create -f vendor/k8s.io/apiextensions-apiserver/artifacts/customresource-01/noxu-resource-definition.yaml 
customresourcedefinition "noxus.mygroup.example.com" created
[deads@deads-dev-01 kubernetes]$ kubectl create -f vendor/k8s.io/apiextensions-apiserver/artifacts/customresource-01/noxu.yaml 
noxu "alfa-noxu" created
[deads@deads-dev-01 kubernetes]$ kubectl delete --cascade=false noxu --all
noxu "alfa-noxu" deleted
(reverse-i-search)`': ^C
[deads@deads-dev-01 kubernetes]$ kubectl get noxu -o yaml
apiVersion: v1
items:
- apiVersion: mygroup.example.com/v1alpha1
  kind: Noxu
  metadata:
    clusterName: ""
    creationTimestamp: 2017-06-13T16:59:29Z
    deletionGracePeriodSeconds: 0
    deletionTimestamp: 2017-06-13T16:59:43Z
    finalizers:
    - orphan
    name: alfa-noxu
    namespace: default
    resourceVersion: "316"
    selfLink: /apis/mygroup.example.com/v1alpha1/namespaces/default/noxus/alfa-noxu
    uid: aa1194f4-5059-11e7-8b44-28d2447dc82b
  spec:
    key: value
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by enisoc** [comment]
Hm for me the last step returns an empty list (delete worked). GC is on by default right? I'm using `local-up-cluster.sh` on master HEAD (prior to your fix merging):

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:44:27Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.0.608+4d31eca42d2390", GitCommit:"4d31eca42d23902a41bbefa7d466c5d0a33cff44", GitTreeState:"clean", BuildDate:"2017-06-13T16:11:29Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by deads2k** [comment]
> Hm for me the last step returns an empty list (delete worked). GC is on by default right? I'm using local-up-cluster.sh on master HEAD (prior to your fix merging):

Your client version looks old.  Try building on the latest.  As I recall, I fixed a cascade bug in 1.7.

**Comment by enisoc** [comment]
Ah ok. I can repro now on a kubectl built from HEAD. It's always fun when fixing bugs uncovers other previously hidden bugs. I'll add a test case for it after your fix merges.
